include/private/net/libssl.hpp

```diff
@@ -0,0 +1,331 @@
+        --------------------
+
+        1.  libssl uses reference counting[1], meaning that the `SSL_CTX *`
+        will be alive as long as there's a `SSL *` using it.
+
+        2. this class does not take any measure to prevent access from multiple
+        threads even though it is not generally wise to assume that OpenSSL
+        will do what you expect on a MT environment [2].
+
+        The policy for managing `Context`s in a MT environment should instead
+        be enforced by the `Cache` implementation (below).
+
+        3. since it manages a C pointer, this class MUST be non-copyable
+        and non-movable to prevent multiple-free bugs.
+
+        Links
+        -----
+
+        .. [1] https://wiki.openssl.org/index.php/Manual:SSL_CTX_free(3)
+        .. [2] https://github.com/openssl/openssl/issues/2165
+    */
+
+    /*!
+        \brief Creates a `SSL *` from the wrapped `SSL_CTX *`.
+
+        \param hostname Hostname for which to create the `SSL *`. This would be
+        the hostname for which you want to use SNI.
+
+        \param logger The logger to use to emit log messages.
+
+        \return An error on failure (the returned object will evaluate to false
+        in a boolean context) and a `SSL *` on success.
+    */
+    template <MK_MOCK(SSL_new)>
+    ErrorOr<SSL *> get_client_ssl(std::string hostname, Var<Logger> logger) {
+        assert(ctx_ != nullptr);
+        SSL *ssl = SSL_new(ctx_);
+        if (ssl == nullptr) {
+            logger->warn("ssl: SSL_new failed");
+            return SslNewError();
+        }
+        SSL_set_tlsext_host_name(ssl, hostname.c_str());
+        return ssl;
+    }
+
+    /// Destructor.
+    ~Context() {
+        assert(ctx_ != nullptr);
+        SSL_CTX_free(ctx_);
+    }
+
+    /*!
+        \brief Factory that creates a shared `SSL_CTX *` wrapper.
+
+        \param ca_bundle_path Path to the CA bundle to be used to verify
+        the servers we're connecting to. If it is empty and we're linking
+        with libressl, we will use the builtin CA bundle that is shipped
+        along with measurement-kit. Otherwise, we will fail.
+
+        \param logger Logger for printing log messages.
+
+        \return An error on failure (the returned object will evaluate to
+        false in a boolean context) or a `Var` shared pointer wrapping a
+        Context instance on success.
+    */
+    template <MK_MOCK(SSL_CTX_new), MK_MOCK(SSL_CTX_load_verify_locations)
+#if (defined LIBRESSL_VERSION_NUMBER && LIBRESSL_VERSION_NUMBER >= 0x2010400fL)
+                                          ,
+              MK_MOCK(SSL_CTX_load_verify_mem)
+#endif
+              >
+    static ErrorOr<Var<Context>> make(std::string ca_bundle_path,
+                                      Var<Logger> logger) {
+        // Implementation note: we need to get the library singleton as early
+        // as here to make sure OpenSSL internals are initialized.
+        auto library = Library::global();
```

Comment:
@hellais reading again my diff, I started wondering whether it could happen that a compiler optimizes away the ``library` variable, since it is not used. We actually only get the singleton for its side effect; i.e. initializing the SSL library.

This pattern of creating a singleton only for its side effect is also used in other places in the library.

I have not worried about this, so far, because compilers I use do not complain about the variable being unused, and because things seems to be still working well.

What is your opinion on that?

Should I refactor this and the other places where this pattern is used, so to avoid this potential problem?

To help understanding the impact:

- if the SSL library is not properly initialized, depending on its version, it will not work (i.e. all calls will fail)

- we use the pattern also for libevent, to make sure locking is initialized; in such case, if the call is optimized away, libevent objects would not be thread safe (which is potentially a catastrophe).